underfs/cephfs/src/main/java/alluxio/underfs/cephfs/CephFSUnderFileSystem.java

```diff
@@ -0,0 +1,792 @@
+   *
+   * @param ufsUri the {@link AlluxioURI} for this UFS
+   * @param conf the configuration for Hadoop
+   * @return a new CephFS {@link UnderFileSystem} instance
+   */
+  public static CephFSUnderFileSystem createInstance(
+      AlluxioURI ufsUri, UnderFileSystemConfiguration conf) throws IOException {
+    LOG.info("CephFS URI: {}", ufsUri.toString());
+
+    /*
+     * Create mount with auth user id
+     */
+    String userId = conf.get(PropertyKey.UNDERFS_CEPHFS_AUTH_ID);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_AUTH_ID, userId);
+    if (userId != null && userId.isEmpty()) {
+      userId = null;
+    }
+    CephMount mount = new CephMount(userId);
+
+    /*
+     * Load a configuration file if specified
+     */
+    String configfile = conf.get(PropertyKey.UNDERFS_CEPHFS_CONF_FILE);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_CONF_FILE, configfile);
+    if (configfile != null && !configfile.isEmpty()) {
+      mount.conf_read_file(configfile);
+    }
+
+    /*
+     * Parse and set Ceph configuration options
+     */
+    String configopts = conf.get(PropertyKey.UNDERFS_CEPHFS_CONF_OPTS);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_CONF_OPTS, configopts);
+    if (configopts != null && !configopts.isEmpty()) {
+      String[] options = configopts.split(";");
+      for (String option : options) {
+        String[] keyval = option.split("=");
+        if (keyval.length != 2) {
+          throw new IllegalArgumentException("Invalid Ceph option: " + option);
+        }
+        String k = keyval[0];
+        String v = keyval[1];
+        try {
+          mount.conf_set(k, v);
+        } catch (Exception e) {
+          throw new IOException("Error setting Ceph option " + k + " = " + v);
+        }
+      }
+    }
+
+    /*
+     * Set auth key
+     */
+    String key = conf.get(PropertyKey.UNDERFS_CEPHFS_AUTH_KEY);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_AUTH_KEY, key);
+    if (key != null && !key.isEmpty()) {
+      mount.conf_set("key", key);
+    }
+
+    /*
+     * Set auth keyfile
+     */
+    String keyfile = conf.get(PropertyKey.UNDERFS_CEPHFS_AUTH_KEYFILE);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_AUTH_KEYFILE, keyfile);
+    if (keyfile != null && !keyfile.isEmpty()) {
+      mount.conf_set("keyfile", keyfile);
+    }
+
+    /*
+     * Set auth keyring
+     */
+    String keyring = conf.get(PropertyKey.UNDERFS_CEPHFS_AUTH_KEYRING);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_AUTH_KEYRING, keyring);
+    if (keyring != null && !keyring.isEmpty()) {
+      mount.conf_set("keyring", keyring);
+    }
+
+    /*
+     * Set monitor hosts
+     */
+    String monHost;
+    monHost = conf.get(PropertyKey.UNDERFS_CEPHFS_MON_HOST);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_MON_HOST, monHost);
+    if (monHost == null || monHost.isEmpty()) {
```

Comment:
Do you think the we should respect the ufsUri first? In my opinion, if the ufsUri specify a concrete host and port, it would be better to use it as monitor host. If ufsUri is an invalid value, we can get the monhost from configuration. Any thoughts?

Reply:
> Do you think the we should respect the ufsUri first? In my opinion, if the ufsUri specify a concrete host and port, it would be better to use it as monitor host. If ufsUri is an invalid value, we can get the monhost from configuration. Any thoughts?

It is ok to use ufsUri first